source/common/stats/thread_local_store.cc

```diff
@@ -257,19 +254,49 @@ void ThreadLocalStoreImpl::releaseScopeCrossThread(ScopeImpl* scope) {
   if (!shutting_down_ && main_thread_dispatcher_) {
     const uint64_t scope_id = scope->scope_id_;
     lock.release();
+
+    // TODO(jmarantz): consider batching all the scope IDs that should be
+    // cleared from TLS caches to reduce bursts of runOnAllThreads on a large
+    // config update. See the pattern below used for histograms.
     main_thread_dispatcher_->post([this, central_cache, scope_id]() {
       sync_.syncPoint(MainDispatcherCleanupSync);
       clearScopeFromCaches(scope_id, central_cache);
     });
   }
 }
 
+void ThreadLocalStoreImpl::releaseHistogramCrossThread(uint64_t histogram_id) {
+  // This can happen from any thread. We post() back to the main thread which will initiate the
+  // cache flush operation.
+  if (!shutting_down_ && main_thread_dispatcher_) {
+    // It's possible that many different histograms will be deleted at the same
+    // time, before the main thread gets a chance to run
+    // clearHistogramsFromCaches. If a new histogram is deleted before that
+    // post runs, we add it to our list of histograms to clear, and there's no
+    // need to issue another post.
+    bool need_post = false;
+    {
+      Thread::LockGuard lock(hist_mutex_);
+      need_post = histograms_to_cleanup_.empty();
+      histograms_to_cleanup_.push_back(histogram_id);
+    }
+    if (need_post) {
+      main_thread_dispatcher_->post([this]() { clearHistogramsFromCaches(); });
+    }
+  }
+}
+
 ThreadLocalStoreImpl::TlsCacheEntry&
 ThreadLocalStoreImpl::TlsCache::insertScope(uint64_t scope_id) {
   return scope_cache_[scope_id];
 }
 
 void ThreadLocalStoreImpl::TlsCache::eraseScope(uint64_t scope_id) { scope_cache_.erase(scope_id); }
+void ThreadLocalStoreImpl::TlsCache::eraseHistograms(const std::vector<uint64_t>& histograms) {
+  for (uint64_t histogram_id : histograms) {
+    histogram_cache_.erase(histogram_id);
```

Comment:
Can we assert that we erased something? Should this never not be true?

Reply:
I think in this scenario we can't do that. There is no guarantee that every thread will have lazily cached a TlsHistogram object for every histogram.

Reply:
added comment explaining this.